### PR TITLE
Dev container Templates & distribution

### PR DESCRIPTION
**Pull request by samruddhikhandale**
The proposal is an idea of how to author and distribute Templates. This follows a very similar mechanism as of dev container [Features](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md)

### CODE REVIEW

File: proposals/devcontainer-templates-distribution.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+│   │   ├── devcontainer-template.json
+│   │   ├── .devcontainer
+│   │       ├── devcontainer.json
+│   │       ├── docker-compose.yml
+│   │       ├── Dockerfile
+│   │       └── ...
+│   │   ├── ...
+```
+
+...where `src` is a directory containing a sub-folder with the name of the Template (e.g. `src/dotnet` or `src/docker-from-docker`) with at least a file named `devcontainer-template.json` that contains the Template metadata, and a `.devcontainer` that the supporting tools will drop into an existing project or folder.
```
Comment (chrmarti) [inline proposals/devcontainer-templates-distribution.md:49]: Do we also want to support the `.devcontainer.json` file in the workspace root folder?

Comment (chrmarti) [reply proposals/devcontainer-templates-distribution.md:49]: Also: Probably need to change `.devcontainer` to `.devcontainer/devcontainer.json` in the text to fully make sense.

File: proposals/devcontainer-templates.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      "default": "value1"
+    }
+  }
+}
+```
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| `optionId` | string | ID of the option that is converted into an all-caps and snake case build argument with the selected value in it. |
+| `optionId.type` | string | Type of the option. Valid types are currently: `boolean`, `string` |
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:63]: Build args are always strings (similar to env variables).

File: proposals/devcontainer-templates.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| `id` | string | ID of the Template. The `id` should be unique in the context of the repository/published package where the Template exists and must match the name of the directory where the `devcontainer-template.json` resides. |
+| `version` | string | The semantic version of the Template. |
+| `name` | string | Name of the Template. |
+| `description` | string | Description of the Template. |
+| `documentationURL` | string | Url that points to the documentation of the Template. |
+| `licenseURL` | string | Url that points to the license of the Template. |
+| `type` | string | Type of the dev container (image, dockerfile, dockerCompose) created by the Template. |
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:36]: How is this used?

File: proposals/devcontainer-templates.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+| `image` | string | Name of an image in a container registry ([DockerHub](https://hub.docker.com/), [GitHub Container Registry](https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages), [Azure Container Registry](https://azure.microsoft.com/en-us/products/container-registry/)) that the [Template supporting services / tools](#template-supporting-tools-and-services) should use to populate list of image variants. |
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:37]: Is this for the UI? Can `options` be used for this instead of the additional `image` property?

File: proposals/devcontainer-templates.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+| `features` | object | List of [Feature id](devcontainer-features.md/#referencing-a-feature) that the [Template supporting services / tools](#template-supporting-tools-and-services) should use to populate different configuration options. |
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:38]: Is this for showing a list to the user to pick from? Would it include the major version?

**Review by chrmarti — CHANGES_REQUESTED** [review]
Looks great overall. Left a few questions.

File: proposals/devcontainer-templates-distribution.md (hunk 1 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates-distribution.md]: > Do we also want to support the .devcontainer.json file in the workspace root folder?
Yep, missed that. Added

File: proposals/devcontainer-templates.md (hunk 3 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md:32]: Currently, the `definitions` have this `type` property (`singleContainer` and `dockerCompose`) which is used by the UI to determine how to start or stop the container and configure few other things. Hence, added it here. 

Also, as per today's discussion, this will also be useful for listing Templates and people can sort with the `type`

File: proposals/devcontainer-templates.md (hunk 4 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Removed as per the discussion and added to `options` instead.

File: proposals/devcontainer-templates.md (hunk 5 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Removed as per the discussion and added to options instead.

File: proposals/devcontainer-templates.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+A Template is a self contained entity in a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.
+
+```
++-- template
+|    +-- devcontainer-template.json
+|    +-- .devcontainer.json
+|    +-- (other files)
+```
+
+## devcontainer-template.json properties
```
Comment (samruddhikhandale) [inline proposals/devcontainer-templates.md:18]: @Chuxel  For the `publisher` info which is needed by the UI -  I believe we will have a similar index for Templates as we have for [Features](https://github.com/devcontainers/devcontainers.github.io/blob/gh-pages/_data/collection-index.yml#L2) and this index will be used by the tooling to display list of Templates. We can then make use of `maintainer` property to display to `publisher/maintainer` name wherever needed.  

 Adding `publisher` prop to `devcontainer-template.json` feels a little unsafe as any author can claim to be someone else and we currently don't have identity checking mechanism in place. At least, with the search index, we will manually look at the PR created by the author (to add his Templates to crawler) and confirm if the maintainer prop matches with  other details.

File: proposals/devcontainer-templates-distribution.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+A user can add the Templates in to their projects as defined by the [Templates Supporting Tools](./devcontainer-templates.md/#adding-a-template-to-a-project-or-codespace).
+
+### OCI Registry
+
+An OCI registry that implements the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec) serves as the primary distribution mechanism for Templates.
+
+Each packaged Template is pushed to the registry following the naming convention `<registry>/<namespace>/<id>[:latest]`.
+
+> **Note:** The `namespace` is a unique indentifier for the collection of templates and must be different than the collection of [Features](./devcontainer-features.md). There are no strict rules for the `namespace`; however, one pattern is to set `namespace` equal to source repository's `<owner>/<repo>`.
```
Comment (joshspicer) [inline proposals/devcontainer-templates-distribution.md:86]:
```suggestion
> **Note:** The `namespace` is a unique identifier for the collection of templates and must be different than the collection of [Features](./devcontainer-features.md). There are no strict rules for the `namespace`; however, one pattern is to set `namespace` equal to source repository's `<owner>/<repo>`. 
```

File: proposals/devcontainer-templates.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
+# Dev Container Templates reference
+
+Development container "Templates" are self-contained, shareable units of development container configuration. The name comes from the idea that adding a Template to your project helps it get up and running with a containerized environment. The Templates describe the appropriate container image, [Features](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md), runtime arguments for starting the container, and VS Code extensions that should be installed. Each provides a container configuration file ([`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson)) and other needed files that you can drop into any existing folder as a starting point for containerizing your project.
```
Comment (joshspicer) [inline proposals/devcontainer-templates.md:3]:
```suggestion
Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project.  A [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.

Configuration is placed in a [`.devcontainer/devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson). The `devcontainer.json` can reference other files within the template, such as a `Dockerfile` or `docker-compose.yml`. 

A template can also provide additional example source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).

```

File: proposals/devcontainer-templates.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+Template metadata is captured by a `devcontainer-template.json` file in the root folder of the Template.
+
+## Folder structure 
+
+A Template is a self contained entity in a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.
```
Comment (joshspicer) [inline proposals/devcontainer-templates.md:9]:
```suggestion
A single Template is a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.
```

File: proposals/devcontainer-templates.md (hunk 3 above)
Comment (joshspicer) [reply proposals/devcontainer-templates.md:31]: Perhaps the tooling could add some of these attributes.  I imagine from some `devcontainer template package` command, it would not be too difficult to programatically add several of these attributes (and ensures consistency/doesn't require action from the author).

File: proposals/devcontainer-templates.md (hunk 8 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Thanks for the edit. looks good, added with little tweaks

File: proposals/devcontainer-templates.md (hunk 3 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md:33]: Just wondering what if the tooling mistags some times ?

Example - lets say a Template contains all of three files (.devcontainer.json, dockerfile and dockercompose), the Template author writes code in such a way that the dev container solely depends on an image and has only added boiler plate code for `dockerfile` and `dockerCompose` or samples of how to configure using them. If the tooling simply determines the value based on the file presence of some other attributes, I wonder it might mis configure sometimes 🤔

File: proposals/devcontainer-templates-distribution.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+### OCI Registry
+
+An OCI registry that implements the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec) serves as the primary distribution mechanism for Templates.
+
+Each packaged Template is pushed to the registry following the naming convention `<registry>/<namespace>/<id>[:latest]`.
+
+> **Note:** The `namespace` is a unique identifier for the collection of templates and must be different than the collection of [Features](./devcontainer-features.md). There are no strict rules for the `namespace`; however, one pattern is to set `namespace` equal to source repository's `<owner>/<repo>`. 
+
+A custom media type `application/vnd.devcontainers` and `application/vnd.devcontainers.layer.v1+tar` are used as demonstrated below.
```
Comment (chrmarti) [inline proposals/devcontainer-templates-distribution.md:88]: @joshspicer Would it make sense to use a different media type for templates and features to distinguish the two? Does OCI's artifact spec have a recommendation?

File: proposals/devcontainer-templates.md [hunk 11]
```diff
@@ -0,0 +1,192 @@
+# Dev Container Templates reference
+
+Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project and a [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.
+
+Configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson). Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files within the template, such as a `Dockerfile` or `docker-compose.yml`. A template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:5]: Can a `.devcontainer.json` also reference other files? (I thought so, but the wording seems to suggest otherwise.)

Prepend "The" to "configuration"?

File: proposals/devcontainer-templates.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+and it has the following `.devcontainer.json` file:
+
+```json
+{
+	"name": "Java",
+	"image": "mcr.microsoft.com/devcontainers/java:0-{{imageVariant}}",
+	"features": {
+		"ghcr.io/devcontainers/features/node:1": {
+			"version": "{{nodeVersion}}",
```
Comment (chrmarti) [inline proposals/devcontainer-templates.md:156]: We are using `${localEnv:<env var name>}` and similar variables at runtime. Should we use `${templateOption:<option id>}` (e.g., `${templateOption:nodeVersion}` here) for consistency?

**Review by chrmarti — CHANGES_REQUESTED** [review]
Great progress. Left a few more questions.

File: proposals/devcontainer-templates-distribution.md (hunk 10 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates-distribution.md:88]: Thinking of having a `label` to distinguish them, if that makes sense. 

> "labels": {
>     "com.github.package.type": "devcontainer_template"
>   }

For Features we already have it as `"com.github.package.type": "devcontainer_feature"`

File: proposals/devcontainer-templates-distribution.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+│   │   ├── ...
+```
+
+...where `src` is a directory containing a sub-folder with the name of the Template (e.g. `src/dotnet` or `src/docker-from-docker`) with at least a file named `devcontainer-template.json` that contains the Template metadata, and a `.devcontainer.json` (or `.devcontainer/devcontainer.json`) that the supporting tools will drop into an existing project or folder.
+
+Each sub-directory should be named such that it matches the `id` field of the `devcontainer-template.json`.  Other files can also be included in the Templates's sub-directory, and will be included during the [packaging step](#packaging) alongside the two required files.  Any files that are not part of the Templates's sub-directory (e.g. outside of `src/dotnet`) will not included in the [packaging step](#packaging).
+
+## Versioning
+
+Each Template is published with only the `latest` tag. Tooling that handles releasing Templates will republish the `latest` tag every time a new release is created.
```
Comment (Chuxel) [inline proposals/devcontainer-templates-distribution.md:53]: Hmmm. Not sure we landed on only publishing latest.  I think what we figured out is that we'd want versioning, but that we also need to factor in the schema version. This would allow us to do automated verion selection based on the schema much like VS Code extensions do today down the road. But you'd need to have some concept of versioning to make that work. Thoughts @chrmarti @alexdima?

File: proposals/devcontainer-templates.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+### Template Supporting Tools and Services
+
+Outlines tools and services that currently support adding Templates to a project.
+
+- [Visual Studio Code Remote - Containers](../docs/specs/supporting-tools.md/#visual-studio-code-remote---containers)
+- [GitHub Codespaces](../docs/specs/supporting-tools.md/#github-codespaces)
+
+## Adding a Template to a project or codespace
+
```
Comment (Chuxel) [inline proposals/devcontainer-templates.md:76]: Since this section is tool / service specific, I'm not sure it belongs in the spec as much as it does documentation for VS Code and Codespaces. It could be added to tool specific steps.  A CLI-based example would be better if there's a need to have something here.

File: proposals/devcontainer-templates.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+| `optionId.type` | string | Type of the option. Valid types are currently: `boolean`, `string` |
+| `optionId.description` | string | Description for the option. |
+| `optionId.proposals` | array | A list of suggested string values. Free-form values **are** allowed. Omit when using `optionId.enum`. |
+| `optionId.enum` | array | A strict list of allowed string values. Free-form values are **not** allowed. Omit when using `optionId.proposals`. |
+| `optionId.default` | string | Default value for the option. |
+| `optionId.replaceIn` | array | List of file paths which the supporting tool should use to perform the string replacement of `optionId` with the selected value. The provided path is always relative to the sub-directory of the Template. |
+
+> `Note`: The `options` must be unique for every `devcontainer-template.json`
+
+### Template Supporting Tools and Services
```
Comment (Chuxel) [inline proposals/devcontainer-templates.md:68]: As with the next section, I'm not sure we need to reference this as a part of the proposal as much as it should be in docs for VS Code and Codespaces.  We could, however, add this into the spec's "implementing tools" doc.  

You could refactor this to be "example usage" if you think it's important, but it might be better just to omit and add the implementing tools part when we move it into containers.dev.

File: proposals/devcontainer-templates.md (hunk 3 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Makes sense. Removed `type`

File: proposals/devcontainer-templates.md (hunk 6 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md:20]: Agreed, added `publisher` 👍

File: proposals/devcontainer-templates.md (hunk 11 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Good catch, updated wordings.

File: proposals/devcontainer-templates.md (hunk 12 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Nice, updated!

File: proposals/devcontainer-templates.md (hunk 14 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Makes sense, removed!

File: proposals/devcontainer-templates.md (hunk 15 above)
Comment (samruddhikhandale) [reply proposals/devcontainer-templates.md]: Makes sense, removed. 👍

File: proposals/devcontainer-templates-distribution.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+This specification defines a pattern where community members and organizations can author and self-publish [Dev Container Templates](./devcontainer-templates.md). 
+
+Goals include:
+
+- For Template authors, create a "self-service" way to publish a Template, either publicly or privately, that is not centrally controlled.
+- Provide the ability to standardize publishing such that supporting tools may implement their own mechanism to aid Template discoverability as they see fit.
+
+## Source code
+
+Templates source code is stored in a git repository.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:12]: Potential nit?
```suggestion
A Template's source code is stored in a git repository.
```

File: proposals/devcontainer-templates-distribution.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+For ease of authorship and maintenance, [1..n] Templates can share a single git repository. This set of Templates is referred to as a "collection," and will share the same [`devcontainer-collection.json`](#devcontainer-collection.json) file and "namespace" (eg. `<owner>/<repo>`).
+
+> **Note:** Templates and [Features](./devcontainer-features.md) should be placed in different git repositories. 
+
+Source code for the set follows the example file structure below:
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:18]:
```suggestion
Source code for a set of Templates follows the example file structure below:
```

File: proposals/devcontainer-templates-distribution.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+### devcontainer-collection.json
+
+The `devcontainer-collection.json` is an auto-generated metadata file.
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| `sourceInformation` | object | Metadata from the implementing packaging tool. |
+| `templates` | array | The list of Templates that are contained in this collection.|
+
+Each templates's `devcontainer-template.json` metadata file is appended into the `templates` top-level array.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:72]: Consistency and spelling nit:
```suggestion
Each Template's `devcontainer-template.json` metadata file is appended into the `templates` top-level array.
```

File: proposals/devcontainer-templates-distribution.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+## Distribution
+
+There are several supported ways to distribute templates.  Distribution is handled by the implementing packaging tool.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:76]:
```suggestion
There are several supported ways to distribute Templates.  Distribution is handled by the implementing packaging tool.
```

File: proposals/devcontainer-templates-distribution.md [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+A user can add the Templates in to their projects as defined by the [Templates Supporting Tools](./devcontainer-templates.md/#adding-a-template-to-a-project-or-codespace).
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:78]:
```suggestion
A user can add Templates in to their projects as defined by the [Templates Supporting Tools](./devcontainer-templates.md/#adding-a-template-to-a-project-or-codespace).
```

Unless we're referring to a specific set of Templates (which would thus be "The" Templates?)

File: proposals/devcontainer-templates-distribution.md [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+### OCI Registry
+
+An OCI registry that implements the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec) serves as the primary distribution mechanism for Templates.
+
+Each packaged Template is pushed to the registry following the naming convention `<registry>/<namespace>/<id>[:latest]`.
+
+> **Note:** The `namespace` is a unique identifier for the collection of templates and must be different than the collection of [Features](./devcontainer-features.md). There are no strict rules for the `namespace`; however, one pattern is to set `namespace` equal to source repository's `<owner>/<repo>`.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:86]:
```suggestion
> **Note:** The `namespace` is a unique identifier for the collection of Templates and must be different than the collection of [Features](./devcontainer-features.md). There are no strict rules for the `namespace`; however, one pattern is to set `namespace` equal to source repository's `<owner>/<repo>`. 
```

File: proposals/devcontainer-templates-distribution.md [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+
+ARTIFACT_PATH=devcontainer-template-go.tgz
+
+oras push ${REGISTRY}/${NAMESPACE}/${TEMPLATE}:latest \
+        --manifest-config /dev/null:application/vnd.devcontainers \
+                ./${ARTIFACT_PATH}:application/vnd.devcontainers.layer.v1+tar
+
+```
+
+The "namespace" is the globally identifiable name for the collection of templates. (eg: `owner/repo` for the source code's git repository).
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates-distribution.md:108]:
```suggestion
The "namespace" is the globally identifiable name for the collection of Templates. (eg: `owner/repo` for the source code's git repository).
```

File: proposals/devcontainer-templates.md [hunk 23]
```diff
@@ -0,0 +1,177 @@
+# Dev Container Templates reference
+
+Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project and a [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:3]:
```suggestion
Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project, and a [supporting tool](https://containers.dev/supporting) will use the configuration from the Template to build a development container.
```

File: proposals/devcontainer-templates.md [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+
+The configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson) which can also reference other files within the Template. Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files, such as a `Dockerfile` or `docker-compose.yml`. A template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:5]:
```suggestion
The configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson) which can also reference other files within the Template. Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files, such as a `Dockerfile` or `docker-compose.yml`. A Template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).
```

File: proposals/devcontainer-templates.md [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+
+The `id` format (`<oci-registry>/<namespace>/<template>[:latest]`) dictates how a [supporting tool](../docs/specs/supporting-tools.md#supporting-tools-and-services) will locate and download a given Template from an OCI registry (For example - `ghcr.io/user/repo/go:latest`). The registry must implement the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec). Some implementors can be [found here](https://oras.land/implementors/). 
+
+## Versioning
+
+Each Template is published with only the `latest` tag. Tooling that handles releasing Templates will republish the `latest` tag every time a new release is created.
+
+## Release
+
+_For information on distribution Templates, see [devcontainer-templates-distribution.md](./devcontainer-templates-distribution.md)._
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:78]:
```suggestion
_For information on distributing Templates, see [devcontainer-templates-distribution.md](./devcontainer-templates-distribution.md)._
```

File: proposals/devcontainer-templates.md [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+
+### Option Resolution
+
+A templates's `options` is used by a supporting tool to prompt for different configuration options. A supporting tool will parse the `options` object provided by the user. If a value is selected for a Template, it will be replaced in the provided `replaceIn` files.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:82]:
```suggestion
A Template's `options` property is used by a supporting tool to prompt for different configuration options. A supporting tool will parse the `options` object provided by the user. If a value is selected for a Template, it will be replaced in the provided `replaceIn` files.
```

File: proposals/devcontainer-templates.md [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+
+### Option resolution example
+
+Consider a `java` template with the following folder structure:
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:86]:
```suggestion
Consider a `java` Template with the following folder structure:
```

File: proposals/devcontainer-templates.md [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+		"ghcr.io/devcontainers/features/node:1": {
+			"version": "${templateOption:nodeVersion}",
+      "installMaven": "${templateOption:installMaven}"
+		}
+	},
+//	...
+}
+```
+
+A user tries to add the `java` Template to their project using the [supporting tools](../docs/specs/supporting-tools.md#supporting-tools-and-services) and selects `17-bullseye` when prompted for `"Specify version of Go."` and uses the `default` values for `"Specify version of node, or 'none' to skip node installation."` and `"Install Maven, a management tool for Java."`.
```
Comment (bamurtaugh) [inline proposals/devcontainer-templates.md:149]:
```suggestion
A user tries to add the `java` Template to their project using the [supporting tools](../docs/specs/supporting-tools.md#supporting-tools-and-services) and selects `17-bullseye` when prompted for `"Specify version of Go"` and the `default` values when prompted for `"Specify version of node, or 'none' to skip node installation"` and `"Install Maven, a management tool for Java"`.
```

**Review by bamurtaugh — COMMENTED** [review]
Thanks for the great progress here!

File: proposals/devcontainer-templates-distribution.md (hunk 13 above)
Comment (chrmarti) [reply proposals/devcontainer-templates-distribution.md:53]: #22 would reference only a devcontainer.json as I understand it, not an entire template.

My take-away on template versioning was that it would not solve the schema version compatibility (the same way the package.json's version property doesn't solve the Node.js version compatibility) and that we would look into adding a `"schemaVersion"` (or similar) property (similar to the package.json's engines property) in a separate proposal to not overload this one.

File: proposals/devcontainer-templates.md (hunk 3 above)
Comment (chrmarti) [reply proposals/devcontainer-templates.md]: Sorry, I wasn't clear: I suggest to keep `type` and we can add tooling support later (whenever we get to it). :)

File: proposals/devcontainer-templates-distribution.md (hunk 13 above)
Comment (Chuxel) [reply proposals/devcontainer-templates-distribution.md:53]: I am quite confident that when we expand #22 to supporting remote devcontainer.jsons, customers will want an entire template since this is generally how the request has come across.  We limited #22 to local to start because this proposal was not in place. The request we've heard often is to have the ability to reference a remote template so that when the template updates with non-breaking changes, downstream repositories get the updates without modification. To achieve this, we'd need to support referencing MAJOR or MAJOR.MINOR versions of templates like we do with features.

Also, with the schema version property, I agree that we do not want to overload this proposal, but fundamentally that will require being able to go back to a previous version of a template.

In either case, including the same versioning semantics not only keeps things consistent with features, but lets clients implement as they see fit and we don't have to retroactively go back and update this proposal and how anyone that is already publishing templates is doing so.

We can omit this from the proposal here - but given what we already know, and how features is set up, why would we do that?

Comment (samruddhikhandale) [reply proposals/devcontainer-templates-distribution.md:53]: Writing down my thoughts for semver vs always publishing `latest` (mostly reiterating what Chuck already mentioned)
- Using semantic versioning will make publishing consistent with Features - it will avoid questions from the community of why one logic for Features and other for Templates.
- Given current consumption of definitions/templates by the Remote Containers UX, I agree versioning does not make complete sense and it doesn't sound helpful. Hence, for UX we can always pin the `latest` tag. However, there is a possibility that the community could develop some tool or service which might require pinning specific versions. So it will be nice to have (just in case)
    - Given our current set of maintained Templates, we might not think that we may add breaking changes to them. However, given that the Templates will be published and distributed by community as well, we can't be sure what pattern they might use. 

To sum up, I believe that having semver may not have useful scenarios currently, but it could be nice to have to provide flexibility and adaptability for future scenarios. Also, I don't see a downside of publishing semantic versions, unless I am missing something super obvious 😄

Comment (chrmarti) [reply proposals/devcontainer-templates-distribution.md:53]: I agree it might be useful in the future, I'm fine with adding it now.

**Review by chrmarti — APPROVED** [review]
LGTM!

**Review by Chuxel — APPROVED** [review]

**Comment by eitsupi** [comment]
Hi, thanks for working on this.
Are you planning to create a template for this as well as feature?

**Comment by samruddhikhandale** [comment]
Hi @eitsupi,
Yes, we will create a Template repo as well when self authoring/publishing mechanism is in place.

**Comment by eitsupi** [comment]
Glad to hear it, and I look forward to seeing the devcontainer ecosystem expand even further!
